### PR TITLE
Topic/skl codeloader fix cppcheck

### LINKED ISSUE #637: SKL/KBL loader is broken
cppcheck reports an issue that exposes the fact that this loader has not worked in a very very long time....

````c
static int cl_skl_cldma_setup_bdle(struct snd_sof_dev *sdev,
				   struct snd_dma_buffer *dmab_data,
				   u32 **bdlp, int size, int with_ioc)
{
	u32 *bdl = *bdlp;
	int frags = 0;

	while (size > 0) {
		phys_addr_t addr = virt_to_phys(dmab_data->area +
						(frags * size));

		bdl[0] = cpu_to_le32(lower_32_bits(addr));
		bdl[1] = cpu_to_le32(upper_32_bits(addr));

		bdl[2] = cpu_to_le32(size);

		size -= size; //<<<<< THIS IS BROKEN
		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);

		bdl += 4;
		frags++;
	}

	return frags;
}
````

### PR DESCRIPTION
**Pull request by zhuyingjiang**
fix #637
fix the skl code loader  code for cppcheck
remove size check since one codeloader dma buffer is enough.
remove wait interrupt return value check since using pollmode.

### CODE REVIEW

File: sound/soc/sof/intel/hda-loader-skl.c [hunk 1]
```diff
@@ -106,30 +106,24 @@
...
 
-		size -= size;
-		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
-
-		bdl += 4;
-		frags++;
-	}
-
-	return frags;
+	return;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader-skl.c]: Humm, that seems too good to be true. Why would there be a single contiguous memory? I don't buy this, sorry

File: sound/soc/sof/intel/hda-loader-skl.c [hunk 2]
```diff
@@ -412,13 +411,6 @@ cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
 			cl_skl_cldma_fill_buffer(sdev, bufsize, bufsize,
 						 curr_pos, true, true);
 
-			if (ret < 0) {
-				dev_err(sdev->dev, "error: fw failed to load. 0x%x bytes remaining\n",
-					bytes_left);
-				cl_skl_cldma_stream_run(sdev, false);
-				return ret;
-			}
-
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader-skl.c:428]: can you please rework the cl_skl_cldma_fill_buffer function? You don't need to pass two arguments for the buffer size and copy size.

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 1 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: @plbossart I keep this aligned with SST driver, I think this need a smaller continuous DMA hardware memory? (CLDMA need SNDRV_DMA_TYPE_DEV type continues buffer, stream DMA use SG buffer). Using a smaller DMA buffer and copy from the binary buffer multi times.
There are two solutions:
1. small DMA hardware buffer and copy to it multi times
`(file buffer) +--------+--------+--------+--`
`(memory copy) 1........2........3........4..`
`(DMA buffer ) +========+`

2. big DMA buffer and copy to it once
`(file buffer) +----------------------------+`
`(memory copy) 1.............................`
`(DMA buffer ) +========+========+========+========+`

In SST driver also use the solution 1:
`	ctx->cl_dev.ops.cl_setup_bdle(ctx, &ctx->cl_dev.dmab_data,
			&bdl, ctx->cl_dev.bufsize, 1);`

the 4th parameter of `size` is `ctx->cl_dev.bufsize`
`static void skl_cldma_setup_bdle(struct sst_dsp *ctx,
		struct snd_dma_buffer *dmab_data,
		__le32 **bdlp, int size, int with_ioc)`

then in skl_cldma_setup_bdle
`		size -= ctx->cl_dev.bufsize;`

which solution do you prefer?

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 2 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c:428]: @plbossart As state as above, if using smaller DMA hardware buffer. we need copy multi times.
And the last buffer copy in may be smaller than the DMA hardware buffer, so we need buffer size and copy size(if size left is bigger than the DMA buffer, copy size equals to DMA buffer size, else using the left size). CSME will verify the buffer the same time as it copy in. And will run the firmware if the verify is OK.

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-loader-skl.c]: Please look at the following code, the assignments are done in a loop.

````c
/* Code loader helper APIs */
static void skl_cldma_setup_bdle(struct sst_dsp *ctx,
		struct snd_dma_buffer *dmab_data,
		__le32 **bdlp, int size, int with_ioc)
{
	__le32 *bdl = *bdlp;

	ctx->cl_dev.frags = 0;
	while (size > 0) {
		phys_addr_t addr = virt_to_phys(dmab_data->area +
				(ctx->cl_dev.frags * ctx->cl_dev.bufsize));

		bdl[0] = cpu_to_le32(lower_32_bits(addr));
		bdl[1] = cpu_to_le32(upper_32_bits(addr));

		bdl[2] = cpu_to_le32(ctx->cl_dev.bufsize);

		size -= ctx->cl_dev.bufsize;
		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);

		bdl += 4;
		ctx->cl_dev.frags++;
	}
}
````

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: @plbossart Yes, there is a loop in `skl_cldma_setup_bdle`, but the place where it is called give the `size` (the 4th parameter) as `ctx->cl_dev.bufsize`:
`	ctx->cl_dev.ops.cl_setup_bdle = skl_cldma_setup_bdle;`
`	ctx->cl_dev.ops.cl_setup_bdle(ctx, &ctx->cl_dev.dmab_data,
			&bdl, ctx->cl_dev.bufsize, 1);`
that make the loop loop only one time. The loops actually done is in `skl_cldma_copy_to_buf`.

What I am missing here?

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: @plbossart

Comment (lyakh) [reply sound/soc/sof/intel/hda-loader-skl.c]: @zhuyingjiang I went a couple of steps further - feel free to use my attached diff
[24.gitdiff.txt](https://github.com/thesofproject/linux/files/2911326/24.gitdiff.txt)

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: @lyakh Thanks, update my PR. except that `bdl[3] = cpu_to_le32(0x01);` should be `bdl[3] = cpu_to_le32(0x0);`
@plbossart Is this what you want?

Comment (lyakh) [reply sound/soc/sof/intel/hda-loader-skl.c]: @zhuyingjiang I don't think bdl[3] should be 0. In the original version:
```
bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
```
where size == 0, with_ioc = 1, i.e.
```
bdl[3] = (0 || !1) ? 0 : cpu_to_le32(0x01);
```
further
```
bdl[3] = 0 ? 0 : cpu_to_le32(0x01);
```
i.e.
```
bdl[3] = cpu_to_le32(0x01);
```

Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: @lyakh Oh, thanks, my mistake! updated.

File: sound/soc/sof/intel/hda-loader-skl.c [hunk 3]
```diff
@@ -106,30 +106,20 @@
...
-		size -= size;
-		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
-
-		bdl += 4;
-		frags++;
-	}
-
-	return frags;
+	bdl[3] = cpu_to_le32(0x01);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader-skl.c:127]: can you please add a comment that the DMA uses a single physically-contiguous memory area and that you want an interrupt at the end of the transfer?

File: sound/soc/sof/intel/hda-loader-skl.c [hunk 4]
```diff
@@ -393,7 +393,7 @@ cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
 			 u32 total_size, u32 bufsize)
 {
 	unsigned int bytes_left = total_size;
-	const void *curr_pos = bin;
+	u8 *curr_pos = (u8 *)bin;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader-skl.c]: why lose the const?

File: sound/soc/sof/intel/hda-loader-skl.c [hunk 5]
```diff
@@ -409,13 +408,6 @@ cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
 			cl_skl_cldma_fill_buffer(sdev, bufsize, bufsize,
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-loader-skl.c]: You can remove one of the two arguments which you test for anyways:
unsigned int size = (bufsize == copysize) ? bufsize : copysize;

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 3 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c:127]: OK, done.

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 4 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: OK, have added.

File: sound/soc/sof/intel/hda-loader-skl.c (hunk 5 above)
Comment (zhuyingjiang) [reply sound/soc/sof/intel/hda-loader-skl.c]: OK, done.

**Comment by zhuyingjiang** [comment]
@plbossart have updated.

**Comment by mengdonglin** [comment]
@zhuyingjiang Share the kmesg?  :-)

**Comment by zhuyingjiang** [comment]
[hda-loader-skl.log](https://github.com/thesofproject/linux/files/2935858/hda-loader-skl.log)
The log attached, note we use force_no_codec mode, and if using the SST firmware to test, we need add firmware strip. 
The step "Firmware download successful, booting..." is reached, and run firmware failed because the FW side is not working.

**Comment by plbossart** [comment]
@zhuyingjiang sorry, I don't get this at all. The firmware boot should work even with a cAVS/SST firmware. To me this means you've used the DMA to do something, but that something is not successful so how can we know that the DMA usage and/or the flow works?

**Comment by zhuyingjiang** [comment]
@plbossart 
This is done by poll reading the FW status register HDA_ADSP_FW_STATUS_SKL, after power on and enabled the core, it will come to 0x1(ROM_INIT), Then the driver will using DMA to copy binary into DSP, the CSME will verify the data automatically, after the manifest is copied and verified, the register HDA_ADSP_FW_STATUS_SKL will come to 0x3(MANIFEST_LOADED), then if the whole binary is copied and passed the authentication, the status will go to 0x4 (FW_LOADED). 
To know why SST firmware/SOF firmware cannot run on DSP, (the HDA_ADSP_FW_STATUS_SKL should go to 0xf: FW_START), we need to investigating into the FW code. Liam said the FW team will look into this if they have bandwidth and found it worth to do. Our scope is change the driver and firmware manifest header so the binary can be successfully loaded into the DSP and passed the authentication.

**Comment by plbossart** [comment]
@zhuyingjiang I can understand that maybe the SOF firmware might have format/manifest issues. But
if you are using the SST firmware with no modification, why would the boot fail? Something is not right here.

**Comment by zhuyingjiang** [comment]
@plbossart OK, I will find a workable RVP/NUC with SST driver and firmware and look into it.

**Pull request merged**